codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +530,66 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
+                try {
+                    try {
+                        if (writing) {
+                            throw new IllegalStateException("write is not re-entrant");
+                        }
+                        // Write the portion of the frame.
+                        writing = true;
+                        needFlush |= payload.write(Math.max(0, allowedBytes));
+                    } catch (Exception e) {
+                        // Mark the state as cancelled, we'll clear the pending queue
+                        // via cancel() below.
+                        cancelled = true;
+                    }
+                    if (!cancelled && payload.size() == 0) {
+                        try {
+                            // This frame has been fully written, remove this frame
+                            // and notify the payload. Since we remove this frame
+                            // first, we're guaranteed that its error method will not
+                            // be called when we call cancel.
+                            pendingWriteQueue.remove();
+                            payload.writeComplete();
+                        } catch (Exception e) {
```

Comment:
I don't think you need these inner try/catch blocks ... the outer try/catch/finally should be sufficient.  Maybe we just don't re-throw in all cases (since it looks like you're trying to avoid the rethrow for the write() case)?